ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -735,3 +745,37 @@ waitForSystemToQuiesce st@PeerSelectionState{
                                       }
         }
   | otherwise = GuardedSkip Nothing
+
+jobVerifyPeerSnapshot :: MonadSTM m
+                      => STM m (Maybe LedgerPeerSnapshot)
+                      -> LedgerPeersConsensusInterface m
+                      -> Job () m (Completion m peeraddr peerconn)
+jobVerifyPeerSnapshot readBigLedgerSnapshot
+                      LedgerPeersConsensusInterface {
+                        lpGetLatestSlot,
+                        lpGetLedgerStateJudgement,
+                        lpGetLedgerPeers }
+  = Job job (const (completion Nothing Nothing)) () "jobVerifyPeerSnapshot"
+  where
+    completion queueJob trace = return . Completion $ \st _now ->
+      Decision {
+        decisionTrace = maybeToList trace,
+        decisionState = st,
+        decisionJobs  = maybeToList queueJob }
+
+    job = atomically $ do
+      lsj <- lpGetLedgerStateJudgement
+      case lsj of
+        TooOld -> do
+          maybeLps <- readBigLedgerSnapshot
+          (snapshotSlotNo, snapshotPools) <-
+            case maybeLps of
+              Just (LedgerPeerSnapshot result) -> pure result
+              Nothing                          -> retry
+
+          check . (snapshotSlotNo ==) =<< lpGetLatestSlot
+          result <- (snapshotPools ==) . accumulateBigLedgerStake <$> lpGetLedgerPeers
+          completion Nothing (Just $ TraceVerifyPeerSnapshot result)
```

Comment:
This job should receive the ledger state judgement and not query for it. The peer selection governor has the current state and jobs should be issued according to it. Otherwise it becomes very hard to reason about. Sure there can be races but it shouldn't really matter if you are just verifying the snapshot is read only.

Reply:
You should throw `throwIO BigLedgerPeerSnapshotError` here instead of the `peerSelectionGovernorLoop` body.

Reply:
Looks like if I throw here, it will be caught be the handler which won't do what I need. I'd have to use async exceptions, but I think the flow control is more explicit and clearer in my current implementation. One can tell from reading the governor loop body that this condition can stop the node - similar to how we handle bootstrapPeersTimeout.

Reply:
yes and in the handler you can do some cleanup stuff and rethrow the exception in the, right?